### PR TITLE
BUG-108315 Persist and provide encryption key to GCP create and start instance requests

### PR DESCRIPTION
**Pull request by daszabo**

### CODE REVIEW

File: core/src/main/resources/schema/app/20180802121537_BUG-108315_add_encrypted_attribute_column_to_template.sql [hunk 1]
```diff
@@ -0,0 +1,10 @@
+-- // BUG-108315 add encrypted attribute column to template
+-- Migration SQL that makes the change goes here.
+ALTER TABLE template ADD COLUMN secretAttributes TEXT;
+
+
+-- //@UNDO
+-- SQL to undo the change goes here.
+ALTER TABLE template DROP COLUMN secretAttributes;
```
Comment (doktoric) [inline core/src/main/resources/schema/app/20180802121537_BUG-108315_add_encrypted_attribute_column_to_template.sql]: can you please add an if exist ?

File: core/src/main/resources/schema/app/20180802121537_BUG-108315_add_encrypted_attribute_column_to_template.sql [hunk 2]
```diff
@@ -0,0 +1,10 @@
+-- // BUG-108315 add encrypted attribute column to template
+-- Migration SQL that makes the change goes here.
+ALTER TABLE template ADD COLUMN secretAttributes TEXT;
```
Comment (doktoric) [inline core/src/main/resources/schema/app/20180802121537_BUG-108315_add_encrypted_attribute_column_to_template.sql]: can please add default value to this column

File: core/src/main/java/com/sequenceiq/cloudbreak/converter/v2/TemplateV2RequestToTemplateRequestConverter.java [hunk 3]
```diff
@@ -38,10 +40,11 @@ public TemplateRequest convert(TemplateV2Request source) {
         } else if (source.getAzureParameters() != null) {
             template.getParameters().putAll(getConversionService().convert(source.getAzureParameters(), Map.class));
         } else if (source.getGcpTemlateParameters() != null) {
-            template.getParameters().putAll(getConversionService().convert(source.getGcpTemlateParameters(), Map.class));
+            GcpParameters gcpParameters = Optional.ofNullable(source.getGcpTemlateParameters()).orElse(new GcpParameters());
+            template.getSecretParameters().putAll(getConversionService().convert(gcpParameters, Map.class));
```
Comment (doktoric) [inline core/src/main/java/com/sequenceiq/cloudbreak/converter/v2/TemplateV2RequestToTemplateRequestConverter.java]: json or map is better?

File: core/src/main/java/com/sequenceiq/cloudbreak/converter/TemplateRequestToTemplateConverter.java [hunk 4]
```diff
@@ -56,12 +58,25 @@ public Template convert(TemplateRequest source) {
...
         }
         return template;
     }
 
+    private Function<Map<String, Object>, Json> toJson() {
+        return value -> {
+            try {
+                return new Json(value);
+            } catch (JsonProcessingException e) {
+                throw new BadRequestException("Invalid parameters");
```
Comment (doktoric) [inline core/src/main/java/com/sequenceiq/cloudbreak/converter/TemplateRequestToTemplateConverter.java]: can you please add here a log line? and specify what is invalid ?

File: core/src/main/java/com/sequenceiq/cloudbreak/converter/TemplateRequestToTemplateConverter.java [hunk 5]
```diff
@@ -56,12 +58,25 @@ public Template convert(TemplateRequest source) {
                 throw new BadRequestException("Invalid parameters");
```
Comment (doktoric) [inline core/src/main/java/com/sequenceiq/cloudbreak/converter/TemplateRequestToTemplateConverter.java]: can you please add here a log line? and specify what is invalid ?

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/service/GcpDiskEncryptionService.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        customerEncryptionKey.setRsaEncryptedKey(encode(rsaWrapped));
+        return customerEncryptionKey;
+    }
+
+    private byte[] encrypt(PublicKey publicKey, byte[] data) {
+        try {
+            Cipher cipher = Cipher.getInstance(RSA_CIPHER_TRANSFORMATION);
+            cipher.init(Cipher.ENCRYPT_MODE, publicKey);
+            return cipher.doFinal(data);
+        } catch (NoSuchAlgorithmException e) {
```
Comment (doktoric) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/service/GcpDiskEncryptionService.java]: I think a log line should be great here

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/service/GcpDiskEncryptionService.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        customerEncryptionKey.setRawKey(encode(getEncryptionKeyBytes(encryptionKey)));
+        return customerEncryptionKey;
+    }
+
+    private byte[] getEncryptionKeyBytes(String encryptionKey) {
+        try {
+            MessageDigest digest = MessageDigest.getInstance(HASH_ALGORITHM);
+            return digest.digest(Optional.ofNullable(encryptionKey).orElse("").getBytes(StandardCharsets.UTF_8));
+        } catch (NoSuchAlgorithmException e) {
+            throw new CloudbreakServiceException("Hashing algorithm, " + HASH_ALGORITHM + " not supported.", e);
```
Comment (doktoric) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/service/GcpDiskEncryptionService.java]: can you please add what we supports?

File: core/src/main/java/com/sequenceiq/cloudbreak/converter/v2/TemplateV2RequestToTemplateRequestConverter.java (hunk 3 above)
Comment (daszabo) [reply core/src/main/java/com/sequenceiq/cloudbreak/converter/v2/TemplateV2RequestToTemplateRequestConverter.java]: It seems to be a convention, that we use a map for parameters. Also map has better support for handling non-existing key-value pairs.

**Pull request merged**